RestLookupCall: fix restriction precedence

bschwarzent opened the pull request:
When creating clones of the template lookup call, values predefined in
the model of a RestLookupCall ('restriction' property) should take
precedence over values that are automatically added to
this._restriction. This behavior was not well defined before, which
could have lead to unexpected behavior.
One example was the 'maxRowCount' property which was always taken from
the lookup call, even when a different value was predefined in the
'restriction' object.

Documentation on these precedence rules was added to RestLookupCall.js.
Additionally, a Jasmine test was added asserting these rules.

306914

cyrill-wyss commented on eclipse-scout-core/src/lookup/RestLookupCall.js:
```diff
@@ -42,6 +71,17 @@ export default class RestLookupCall extends LookupCall {
 
+  /**
+   * Adds the given key-value pair to 'this._restriction', but only if there is no predefined
+   * value for this this key in 'this.restriction'. This prevents unintentional overriding of
```
Remove one 'this'

cyrill-wyss commented on eclipse-scout-core/test/lookup/RestLookupCallSpec.js:
```diff
@@ -0,0 +1,327 @@
+    expect(lookupCall.restriction).toBeNull();
+    expect(lookupCall._restriction).toBeNull();
+
+    const cloneAll = lookupCall.cloneForAll();
```
The checks from the different test functions look quite similar. Maybe it would be better to extract them into separate functions?

bschwarzent replied:
2 of the 3 test methods were almost identical, I merged them into one. The third has many small differences. It may be a bit verbose, but at least the test cases are clearly identifiable.

cyrill-wyss commented on eclipse-scout-core/test/lookup/RestLookupCallSpec.js:
```diff
@@ -0,0 +1,327 @@
+  });
+
+  it('applies restrictions in the expected order', () => {
+    // This test asserts that the following precedence rules are applies then creating the final restriction object:
```
The list does not match the one in the lookup call comment.

bschwarzent replied:
Correct, I updated the comment

cyrill-wyss commented on eclipse-scout-core/test/lookup/RestLookupCallSpec.js:
```diff
@@ -0,0 +1,327 @@
+      resourceUrl: 'test-api/dummy',
+      maxRowCount: 777,
+      active: true, // will be ignored
+      myProperty: 'one',
```
I guess this should be ignored too?

bschwarzent replied:
Yes, comment was added

cyrill-wyss commented on eclipse-scout-core/test/lookup/RestLookupCallSpec.js:
```diff
@@ -0,0 +1,327 @@
+      myDefaultProperty: 'three'
+    };
+
+    const expectedDefaultRestriction = {
```
looks a bit strange to have two semantic identical objects => Add a comment to describe the position change in the ajax call restriction

bschwarzent replied:
Semantically, they are not exactly the same, but in this particular case, they contain exactly the same properties (the order is not important). There is no default property left that could be used to show that difference, therefore I removed one of these objects.

paolobazzi commented on eclipse-scout-core/src/lookup/RestLookupCall.js:
```diff
@@ -8,9 +8,38 @@
+ *
+ * Order of precedence (lowest to highest):
+ * 1. Restrictions automatically applied to all clones after their creation.
+ *    These are: 'active' (ALL, TEXT, REC) and 'maxRowCount' (ALL, TEXT, REC)
```
maybe add:  
"1. Restrictions automatically applied to all clones after their creation **by their corresponding cloneFor method**"

bschwarzent replied:
OK

bschwarzent committed:
RestLookupCall: fix restriction precedence

When creating clones of the template RestLookupCall, values predefined
in the model of the lookup call ('restriction' property) should take
precedence over values that are automatically added to
this._restriction in the respective cloneForX method. This behavior was
not well defined before, which could have lead to unexpected behavior.

One example is the 'maxRowCount' property which was always taken from
the lookup call, even when a different value was predefined in the
restriction object.

Documentation for these precedence rules was added to RestLookupCall.js.
Additionally, a Jasmine test was added asserting these rules.

306914
